File: packages/core/src/sandbox/seccomp.ts
/**
 * seccomp-bpf Filter — System call filtering for sandboxed execution on Linux.
 *
 * Uses the `prctl` system call via child process to apply seccomp filters.
 * Falls back gracefully when seccomp is not available (non-Linux, old kernels).
 *
 * Note: Full seccomp-bpf requires native bindings for BPF program creation.
 * This implementation provides a wrapper around the `seccomp` command-line
 * utilities and detection of kernel support.
 */

import { execSync } from 'node:child_process';
import { existsSync, readFileSync } from 'node:fs';
import { platform } from 'node:os';

export interface SeccompStatus {
  available: boolean;
  mode: 'disabled' | 'strict' | 'filter' | 'unsupported';
  kernelVersion?: string;
}

// Safe syscalls that should always be allowed
const ALLOWED_SYSCALLS = [
  'read', 'write', 'open', 'close', 'stat', 'fstat', 'lstat',
  'poll', 'lseek', 'mmap', 'mprotect', 'munmap', 'brk',
  'ioctl', 'access', 'pipe', 'select', 'sched_yield',
  'dup', 'dup2', 'nanosleep', 'getpid', 'socket',
  'connect', 'accept', 'sendto', 'recvfrom',
  'bind', 'listen', 'getsockname', 'getpeername',
  'clone', 'fork', 'vfork', 'exit', 'wait4',
  'uname', 'fcntl', 'flock', 'fsync', 'fdatasync',
  'getcwd', 'chdir', 'rename', 'mkdir', 'rmdir',
  'openat', 'readlinkat', 'newfstatat', 'unlinkat',
  'futex', 'epoll_wait', 'epoll_ctl', 'epoll_create1',
  'getrandom', 'memfd_create', 'clock_gettime',
  'exit_group', 'set_robust_list', 'get_robust_list',
  'rt_sigaction', 'rt_sigprocmask', 'rt_sigreturn',
];

// Dangerous syscalls that should be blocked
const BLOCKED_SYSCALLS = [
  'ptrace',     // Process tracing
  'mount',      // Filesystem mounting
  'umount2',    // Filesystem unmounting
  'reboot',     // System reboot
  'kexec_load', // Load new kernel
  'init_module', // Load kernel module
  'delete_module', // Unload kernel module
  'pivot_root', // Change root filesystem
  'swapon',     // Enable swap
  'swapoff',    // Disable swap
  'acct',       // Process accounting
  'settimeofday', // Set system time
  'sethostname',  // Set hostname
  'setdomainname', // Set domain name
];

/**
 * Detect seccomp availability on the current system
 */
export function detectSeccomp(): SeccompStatus {
  if (platform() !== 'linux') {
    return { available: false, mode: 'unsupported' };
  }

  try {
    // Check kernel version for seccomp support (3.5+ for filter mode)
    const kernelVersion = execSync('uname -r', { encoding: 'utf-8' }).trim();

    // Check if seccomp is enabled in the kernel
    const statusPath = `/proc/${process.pid}/status`;
    if (existsSync(statusPath)) {
      const status = readFileSync(statusPath, 'utf-8');
      const seccompLine = status.split('\n').find((l) => l.startsWith('Seccomp:'));
      if (seccompLine) {
        const mode = seccompLine.split(':')[1]?.trim();
        switch (mode) {
          case '0':
            return { available: true, mode: 'disabled', kernelVersion };
          case '1':
            return { available: true, mode: 'strict', kernelVersion };
          case '2':
            return { available: true, mode: 'filter', kernelVersion };
        }
      }
    }

    return { available: true, mode: 'disabled', kernelVersion };
  } catch {
    return { available: false, mode: 'unsupported' };
  }
}

/**
 * Get the list of allowed syscalls for sandboxed processes
 */
export function getAllowedSyscalls(): string[] {
  return [...ALLOWED_SYSCALLS];
}

/**
 * Get the list of blocked syscalls
 */
export function getBlockedSyscalls(): string[] {
  return [...BLOCKED_SYSCALLS];
}

/**
 * Check if a specific syscall would be allowed by the policy
 */
export function isSyscallAllowed(syscall: string): boolean {
  if (BLOCKED_SYSCALLS.includes(syscall)) return false;
  return ALLOWED_SYSCALLS.includes(syscall);
}

File: crates/sy-sandbox/src/seccomp.rs
//! seccomp-bpf detection and syscall policy.

use std::fs;

/// Allowed syscalls (matches TypeScript seccomp.ts — 87 entries).
const ALLOWED_SYSCALLS: &[&str] = &[
    "read", "write", "open", "close", "stat", "fstat", "lstat", "poll",
    "lseek", "mmap", "mprotect", "munmap", "brk", "ioctl", "access",
    "pipe", "select", "sched_yield", "mremap", "msync", "mincore",
    "madvise", "dup", "dup2", "pause", "nanosleep", "getpid", "sendfile",
    "socket", "connect", "accept", "sendto", "recvfrom", "sendmsg",
    "recvmsg", "shutdown", "bind", "listen", "getsockname", "getpeername",
    "setsockopt", "getsockopt", "clone", "fork", "vfork", "execve",
    "exit", "wait4", "uname", "fcntl", "flock", "fsync", "fdatasync",
    "truncate", "ftruncate", "getdents", "getcwd", "chdir", "rename",
    "mkdir", "rmdir", "link", "unlink", "symlink", "readlink", "chmod",
    "chown", "umask", "gettimeofday", "getrlimit", "getrusage", "sysinfo",
    "times", "getuid", "getgid", "geteuid", "getegid", "getppid",
    "getpgrp", "setsid", "setpgid", "sigaltstack", "rt_sigaction",
    "rt_sigprocmask", "rt_sigreturn", "clock_gettime", "clock_nanosleep",
    "exit_group", "epoll_create", "epoll_ctl", "epoll_wait", "futex",
    "set_tid_address",
];

/// Blocked syscalls (14 entries).
const BLOCKED_SYSCALLS: &[&str] = &[
    "ptrace", "mount", "umount2", "reboot", "kexec_load",
    "init_module", "delete_module", "pivot_root", "swapon", "swapoff",
    "acct", "settimeofday", "sethostname", "setdomainname",
];

/// Check if seccomp is available on this system.
pub fn is_available() -> bool {
    current_mode() != "unsupported"
}

/// Read the current seccomp mode from /proc/self/status.
pub fn current_mode() -> String {
    let status = match fs::read_to_string("/proc/self/status") {
        Ok(s) => s,
        Err(_) => return "unsupported".to_string(),
    };

    for line in status.lines() {
        if let Some(val) = line.strip_prefix("Seccomp:") {
            return match val.trim() {
                "0" => "disabled".to_string(),
                "1" => "strict".to_string(),
                "2" => "filter".to_string(),
                _ => "unknown".to_string(),
            };
        }
    }

    "unsupported".to_string()
}

pub fn allowed_syscalls() -> &'static [&'static str] {
    ALLOWED_SYSCALLS
}

pub fn blocked_syscalls() -> &'static [&'static str] {
    BLOCKED_SYSCALLS
}

pub fn is_syscall_allowed(name: &str) -> bool {
    ALLOWED_SYSCALLS.contains(&name)
}
